File: backend/packages/users/src/services/create.py
import redis
import hashlib
import psycopg2.extras
from psycopg2.extensions import connection
from ..utils.service import BaseService


class CreateService(BaseService):
    _redis_client: redis.Redis
    _postgres_client: connection

    def __init__(self, redis_client: redis.Redis, postgres_client: connection) -> None:
        self._redis_client = redis_client
        self._postgres_client = postgres_client

    def call(self, email: str, name: str, password: str):
        user = None
        password_hash = hashlib.sha512(password.encode("utf-8")).hexdigest()

        with self._postgres_client.cursor(
            cursor_factory=psycopg2.extras.RealDictCursor
        ) as cursor:
            cursor.execute(
                """
                INSERT INTO users (email, name, password)
                VALUES (%s, %s, %s)
                RETURNING *
                """,
                (email, name, password_hash),
            )
            self._postgres_client.commit()
            user = cursor.fetchone()
            del user["password"]

        return {"user": user}

File: backend/packages/projects/src/services/create.ts

import redis from 'redis'
import pg from 'pg'
import aws from 'aws-sdk'
import { BaseService } from '../utils/service'

type Input = {
    connectionId: string
    name: string
    description: string
}

interface CreateService extends BaseService<Input, any> { }

type Self = {
    callbackAPIClient: aws.ApiGatewayManagementApi
    redisClient: redis.RedisClientType
    postgresClient: pg.Pool
}

export const makeCreateService = (callbackAPIClient: aws.ApiGatewayManagementApi, redisClient: redis.RedisClientType, postgresClient: pg.Pool): CreateService => {
    const self: Self = {
        callbackAPIClient,
        redisClient,
        postgresClient
    }

    return {
        call: call(self)
    }
}

const call = (self: Self): CreateService['call'] => async (input) => {
    const connectionId = input.connectionId
    const name = input.name
    const description = input.description

    const rawConnection = await self.redisClient.HGET("connections", connectionId)
    const connection = rawConnection && JSON.parse(rawConnection)
    const user = connection?.user

    if (!user) {
        return {
            reason: 'Not authenticated'
        }
    }

    const userId = user.user_id

    const projects = await self.postgresClient.query(`
        INSERT INTO projects (user_id, name, description)
        VALUES ($1, $2, $3)
        RETURNING *
    `, [userId, name, description])

    const connectionIds = await self.redisClient.SMEMBERS(`users:${userId}`)

    for await (const connectionId of connectionIds) {
        if (connection === input.connectionId) {
            continue
        }

        await self.callbackAPIClient.postToConnection({
            ConnectionId: connectionId,
            Data: JSON.stringify({
                action: "projects-create",
                data: {
                    project: projects.rows[0]
                }
            })
        }).promise()
    }

    return {
        project: projects.rows[0]
    }
}